### PR TITLE
parquet

### PR DESCRIPTION
**Pull request by nofateg**
Main changes:
Simplification oap-template and bug fixes. For this, the implementation of grammar -> render tree -> java class is replaced by grammar -> AST tree -> render tree -> java class

### CODE REVIEW

File: oap-template/src/main/java/oap/template/tree/Exprs.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+package oap.template.tree;
+
+import lombok.ToString;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@ToString
+public class Exprs {
```
Comment (vdiravka) [inline oap-template/src/main/java/oap/template/tree/Exprs.java:33]: Is  this just for printing list of Exprs? Possibly static util method for this purpose?
What is the difference with `Expression`? That one also has `ArrayList<Expr>`

File: oap-template/src/main/java/oap/template/tree/Func.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+package oap.template.tree;
+
+import lombok.ToString;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@ToString
+public class Func {
```
Comment (vdiravka) [inline oap-template/src/main/java/oap/template/tree/Func.java:33]: What functions are available? Do we have function types? Any doc for it?

File: oap-template/src/main/java/oap/template/tree/Math.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package oap.template.tree;
+
+import lombok.ToString;
+
+@ToString
+public class Math {
```
Comment (vdiravka) [inline oap-template/src/main/java/oap/template/tree/Math.java:30]: What is the difference with `Func`?

File: oap-template/src/test/java/oap/template/TestTemplateEnumWithoutDefaultValue.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package oap.template;
+
+public enum TestTemplateEnumWithoutDefaultValue {
```
Comment (vdiravka) [inline oap-template/src/test/java/oap/template/TestTemplateEnumWithoutDefaultValue.java:27]: Same as for above `TestTemplateEnum`

File: pom.xml [hunk 5]
```diff
@@ -44,7 +44,7 @@
     </dependencies>
 
     <properties>
-        <oap.project.version>17.10.31.10</oap.project.version>
+        <oap.project.version>17.10.32.10</oap.project.version>
```
Comment (vdiravka) [inline pom.xml:47]: Just a note: need to plan in separate task updating `OAP` (within parent POM) to 18 version

**Review by vdiravka — CHANGES_REQUESTED** [review]
@nofateg 
Please add PR Description, how OAP template-engine updated.

As for me these changes should be merged to OAP master too
@vovakyrychenko

File: oap-template/src/main/java/oap/template/tree/Math.java (hunk 3 above)
Comment (nofateg) [reply oap-template/src/main/java/oap/template/tree/Math.java:30]: syntax, a + b and func(1)

File: oap-template/src/main/antlr4/TemplateGrammar.g4 [hunk 6]
```diff
@@ -25,37 +27,14 @@ import java.util.Map;
 
 }
 
-template[TemplateType parentType, Map<String,String> aliases] returns [AstRoot rootAst]
-	: elements[parentType, aliases] { $rootAst = new AstRoot($parentType); $rootAst.addChildren($elements.list); } EOF
-	;
-
-elements[TemplateType parentType, Map<String,String> aliases] returns [ArrayList<Ast> list = new ArrayList<>()]
-	: (element[parentType, aliases] { $list.add($element.ast); })*
+elements[Map<String,String> aliases] returns [Elements ret = new Elements()]
```
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/antlr4/TemplateGrammar.g4:30]: no changes

File: oap-template/src/main/antlr4/TemplateGrammarExpression.g4 [hunk 7]
```diff
@@ -9,10 +9,16 @@ options {
 @header {
 package oap.template;
 
+import oap.template.tree.*;
```
Comment (vdiravka) [inline oap-template/src/main/antlr4/TemplateGrammarExpression.g4:12]: Is it possible to place exact imports here?

File: oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammar.interp [hunk 8]
```diff
@@ -27,4 +26,4 @@ expressionContent
...
... 2, 12, 2, 36, 1, 3, 1, 3, 1, 3, 1, 3, 1, 4, 1, 4, 1, 4, 1, 4, 1, 4, 1, 5, 4, 5, 49, 8, 5, 11, 5, 12, 5, 50, 1, 5, 0, 0, 6, 0, 2, 4, 6, 8, 10, 0, 1, 1, 0, 4, 6, 51, 0, 17, 1, 0, 0, 0, 2, 31, 1, 0, 0, 0, 4, 34, 1, 0, 0, 0, 6, 38, 1, 0, 0, 0, 8, 42, 1, 0, 0, 0, 10, 48, 1, 0, 0, 0, 12, 13, 3, 2, 1, 0, 13, 14, 6, 0, -1, 0, 14, 16, 1, 0, 0, 0, 15, 12, 1, 0, 0, 0, 16, 19, 1, 0, 0, 0, 17, 15, 1, 0, 0, 0, 17, 18, 1, 0, 0, 0, 18, 20, 1, 0, 0, 0, 19, 17, 1, 0, 0, 0, 20, 21, 5, 0, 0, 1, 21, 1, 1, 0, 0, 0, 22, 23, 3, 4, 2, 0, 23, 24, 6, 1, -1, 0, 24, 32, 1, 0, 0, 0, 25, 26, 3, 6, 3, 0, 26, 27, 6, 1, -1, 0, 27, 32, 1, 0, 0, 0, 28, 29, 3, 8, 4, 0, 29, 30, 6, 1, -1, 0, 30, 32, 1, 0, 0, 0, 31, 22, 1, 0, 0, 0, 31, 25, 1, 0, 0, 0, 31, 28, 1, 0, 0, 0, 32, 3, 1, 0, 0, 0, 33, 35, 5, 3, 0, 0, 34, 33, 1, 0, 0, 0, 35, 36, 1, 0, 0, 0, 36, 34, 1, 0, 0, 0, 36, 37, 1, 0, 0, 0, 37, 5, 1, 0, 0, 0, 38, 39, 5, 1, 0, 0, 39, 40, 3, 10, 5, 0, 40, 41, 5, 5, 0, 0, 41, 7, 1, 0, 0, 0, 42, 43, 5, 2, 0, 0, 43, 44, 3, 10, 5, 0, 44, 45, 5, 5, 0, 0, 45, 46, 6, 4, -1, 0, 46, 9, 1, 0, 0, 0, 47, 49, 7, 0, 0, 0, 48, 47, 1, 0, 0, 0, 49, 50, 1, 0, 0, 0, 50, 48, 1, 0, 0, 0, 50, 51, 1, 0, 0, 0, 51, 11, 1, 0, 0, 0, 4, 17, 31, 36, 50]
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammar.interp:29]: Could you add new line?

File: oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammar.java [hunk 9]
```diff
@@ -185,28 +135,30 @@ public void exitRule(ParseTreeListener listener) {
 		}
 	}
 
-	public final ElementsContext elements(TemplateType parentType,Map<String,String> aliases) throws RecognitionException {
-		ElementsContext _localctx = new ElementsContext(_ctx, getState(), parentType, aliases);
-		enterRule(_localctx, 2, RULE_elements);
+	public final ElementsContext elements(Map<String,String> aliases) throws RecognitionException {
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammar.java:138]: created separate ticket for it
[OAP-96](https://youtrack.xenoss.io/issue/OAP-96) Spaces in brackets codestyle

File: oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammarBaseListener.java [hunk 10]
```diff
@@ -2,6 +2,8 @@
 
 package oap.template;
 
+import oap.template.tree.*;
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammarBaseListener.java:5]: Given the explanation this is hard to list all needed classes in the `antlr` template

[the remark this answers is not in the document]
Comment (vdiravka) [reply oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammarBaseListener.java:5]: Created separate task to handle it:
[OAP-97](https://youtrack.xenoss.io/issue/OAP-97) Eliminate star * imports

File: oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammarBaseListener.java [hunk 11]
```diff
@@ -19,18 +21,6 @@
  */
 @SuppressWarnings("CheckReturnValue")
 public class TemplateGrammarBaseListener implements TemplateGrammarListener {
-	/**
-	 * {@inheritDoc}
-	 *
-	 * <p>The default implementation does nothing.</p>
-	 */
-	@Override public void enterTemplate(TemplateGrammar.TemplateContext ctx) { }
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammarBaseListener.java:27]: Given explanation this is `antlr` generated code. So related to `antl` grammar templates

File: oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammarExpression.interp [hunk 12]
```diff
@@ -72,9 +72,9 @@ longRule
 function
 functionArgs
 functionArg
-orExps
-exps
-exp
+orExprs
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammarExpression.interp:75]: Given explanation this is antlr generated code. So related to antl grammar templates

File: oap-template/src/test/java/oap/template/TestTemplateEnumWithoutDefaultValue.java (hunk 4 above)
Comment (vdiravka) [reply oap-template/src/test/java/oap/template/TestTemplateEnumWithoutDefaultValue.java:27]: Given the explanation the Test prefix is used for test helper classes and Test suffix for actual test classes.

File: pom.xml (hunk 5 above)
Comment (vdiravka) [reply pom.xml:47]: Separate task:
[OAP-98](https://youtrack.xenoss.io/issue/OAP-98) Update Parent POM and migrate to JDK18 for OAP projects

File: oap-template/src/test/java/oap/template/TestTemplateEnum.java [hunk 13]
```diff
@@ -25,5 +25,5 @@
 package oap.template;
 
 public enum TestTemplateEnum {
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply oap-template/src/test/java/oap/template/TestTemplateEnum.java:27]: Given the explanation the Test prefix is used for test helper classes and Test suffix for actual test classes.

File: oap-template/src/main/antlr4/TemplateGrammarExpression.g4 (hunk 7 above)
Comment (vdiravka) [reply oap-template/src/main/antlr4/TemplateGrammarExpression.g4:12]: [OAP-97](https://youtrack.xenoss.io/issue/OAP-97) Eliminate star * imports

File: .idea/codeStyles/Project.xml [hunk 14]
```diff
@@ -1,45 +0,0 @@
-<component name="ProjectCodeStyleConfiguration">
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply .idea/codeStyles/Project.xml:1]: [OAP-99](https://youtrack.xenoss.io/issue/OAP-99) Add .idea to gitignore for all OAP projects

File: .idea/checkstyle-idea.xml [hunk 15]
```diff
@@ -1,20 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<project version="4">
-  <component name="CheckStyle-IDEA" serialisationVersion="2">
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply .idea/checkstyle-idea.xml:3]: [OAP-99](https://youtrack.xenoss.io/issue/OAP-99) Add .idea to gitignore for all OAP projects

File: oap-template/src/main/antlr4/TemplateGrammar.g4 (hunk 6 above)
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/antlr4/TemplateGrammar.g4:30]: AST tree

File: oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammar.interp (hunk 8 above)
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/java-antlr-generated/oap/template/TemplateGrammar.interp:29]: java-antlr-generated/...

File: oap-template/src/main/java/oap/template/tree/Node.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package oap.template.tree;
+
+public interface Node {
```
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/java/oap/template/tree/Node.java:37]: \+

File: oap-template/src/main/antlr4/TemplateGrammarExpression.g4 [hunk 17]
```diff
@@ -25,45 +31,37 @@ import java.util.Map;
...
-        $ast.addLeafs( () -> getAst($ast.bottom.type, null, false, $defaultValue.ctx != null ? $defaultValue.v : null, null ).top );
-
-        java.util.function.Function<Ast, AstText> printDefautlValueAst = ast -> new AstText($defaultValue.ctx != null ? $defaultValue.v : null );
-        $ast.update( ast -> ast instanceof AstNullable, ast -> ((AstNullable)ast).elseAst = printDefautlValueAst.apply( ast ) );
-        $ast.update( ast -> ast instanceof AstOptional, ast -> ((AstOptional)ast).elseAst = printDefautlValueAst.apply( ast ) );
-
-        if( $comment != null ) {
-            $ast.setTop( new AstComment( parentType, $comment ) );
-        }
+expression returns [Expression ret]
```
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/antlr4/TemplateGrammarExpression.g4:34]: I think there should be a separate documentation on the templates.

File: oap-template/src/main/java/oap/template/LogConfiguration.java [hunk 18]
```diff
@@ -1,237 +0,0 @@
...
-import static oap.template.TemplateAccumulators.STRING;
-import static oap.util.Pair.__;
-
-/**
- * Created by igor.petrenko on 2020-07-15.
- * No longer needed, see the latest logstream
- */
-@Slf4j
-@Deprecated
-public class LogConfiguration extends Configuration {
```
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/java/oap/template/LogConfiguration.java:59]: It is not needed anywhere

File: oap-template/src/main/java/oap/template/Types.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    DATE( 3 ),
+    BOOLEAN( 4 ),
+    BYTE( 5 ),
+    SHORT( 6 ),
+    INTEGER( 7 ),
+    LONG( 8 ),
+    FLOAT( 9 ),
+    DOUBLE( 10 ),
+    STRING( 11 ),
+    LIST( 12 );
```
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/java/oap/template/Types.java:42]: We don't use these types so there is no support

File: oap-template/src/main/java/oap/template/render/FieldType.java [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+import oap.util.Strings;
+import org.apache.commons.lang3.ClassUtils;
+
+import java.lang.reflect.ParameterizedType;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+@EqualsAndHashCode
+public class FieldType {
```
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/java/oap/template/render/FieldType.java:38]: Are not related in any way.

File: oap-template/src/main/java/oap/template/tree/Func.java (hunk 2 above)
Comment (nofateg) [reply oap-template/src/main/java/oap/template/tree/Func.java:33]: \+

File: oap-template/src/main/java/oap/template/tree/Expression.java [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+
+package oap.template.tree;
+
+import lombok.ToString;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@ToString
+public class Expression {
```
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/java/oap/template/tree/Expression.java:33]: @see oap-template/README.md

File: oap-template/src/main/java/oap/template/tree/Exprs.java (hunk 1 above)
Comment (nofateg) [reply oap-template/src/main/java/oap/template/tree/Exprs.java:33]: @see oap-template/README.md

File: oap-template/src/main/java/oap/template/tree/Expression.java [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    public final String comment;
+    public final String castType;
+    public final ArrayList<Exprs> or = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/java/oap/template/tree/Expression.java:36]: @see oap-template/README.md

File: oap-template/src/main/java/oap/template/render/TemplateAstUtils.java [hunk 23]
```diff
@@ -0,0 +1,374 @@
...
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static oap.template.ErrorStrategy.IGNORE;
+
+@Slf4j
+public class TemplateAstUtils {
```
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/java/oap/template/render/TemplateAstUtils.java:61]: @see oap-template/README.md

File: oap-template/src/main/java/oap/template/render/AstRenderPrintValue.java [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+import lombok.ToString;
+import oap.util.Dates;
+import oap.util.Strings;
+import org.apache.commons.lang3.StringUtils;
+import org.joda.time.DateTime;
+
+import java.util.Collection;
+
+@ToString( callSuper = true )
+public class AstRenderPrintValue extends AstRender {
```
[the remark this answers is not in the document]
Comment (nofateg) [reply oap-template/src/main/java/oap/template/render/AstRenderPrintValue.java:37]: Each type in this AST tree creates its own renderer.

**Pull request merged**